File: week20_8_10/BOJ_1325_민지.py
"""
해커 김지민은 잘 알려진 어느 회사를 해킹하려고 한다.
이 회사는 N개의 컴퓨터로 이루어져 있다.
김지민은 귀찮기 때문에, 한 번의 해킹으로 여러 개의 컴퓨터를 해킹 할 수 있는 컴퓨터를 해킹하려고 한다.
이 회사의 컴퓨터는 신뢰하는 관계와, 신뢰하지 않는 관계로 이루어져 있는데,
A가 B를 신뢰하는 경우에는 B를 해킹하면, A도 해킹할 수 있다는 소리다.
이 회사의 컴퓨터의 신뢰하는 관계가 주어졌을 때, 한 번에 가장 많은 컴퓨터를 해킹할 수 있는 컴퓨터의 번호를 출력하는 프로그램을 작성하시오.

입력
첫째 줄에, N과 M이 들어온다.
N은 10,000보다 작거나 같은 자연수, M은 100,000보다 작거나 같은 자연수이다.
둘째 줄부터 M개의 줄에 신뢰하는 관계가 A B와 같은 형식으로 들어오며, "A가 B를 신뢰한다"를 의미한다.
컴퓨터는 1번부터 N번까지 번호가 하나씩 매겨져 있다.

출력
첫째 줄에, 김지민이 한 번에 가장 많은 컴퓨터를 해킹할 수 있는 컴퓨터의 번호를 오름차순으로 출력한다.

변수
N = 컴퓨터의 갯수
M = 신뢰하는 연결의 갯수
AL = 컴퓨터들의 연결 고리 정보를 저장하는 이중 리스트
ex) AL[1] 에는 1번 컴퓨터를 해킹했을때 해킹할 수 있는 다른 컴퓨터들의 리스트가 저장되어있음
"""
from copy import deepcopy

N, M = map(int, input().split())
AL = [[] for _ in range(N+1)]
max_cnt = 0
max_nodes = []

for _ in range(M):
    A, B = map(int, input().split())
    AL[B].append(A)

# 시작점을 1~N까지로 설정해서 각각 DFS를 돈다
for s in range(1, N+1):
    visited = [0] * (N+1)
    visited[s] = 1
    stack = [s]
    cnt = 1
    # DFS 시작
    while stack:
        current = stack.pop()
        for i in AL[current]:
            if not visited[i]:
                visited[i] = 1
                stack.append(i)
                cnt += 1
    # 만약 cnt가 현재 나왔던 것들 중 max라면 max를 교체한다
    if cnt > max_cnt:
        max_cnt = cnt
        max_nodes = [s]
    elif cnt == max_cnt:
        max_nodes.append(s)

print(*max_nodes)

File: week20_8_10/BOJ_1325_민지.java
package week20_8_10;

import java.util.ArrayList;
import java.util.Scanner;

public class BOJ_1325_민지 {
    public static void main(String[] args) {
        // input 받기
        // Java에서 변수 초기화 하는 기본적인 문법: <변수 타입> <변수 이름> = <Instance 초기화>
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt(); // 컴퓨터의 갯수
        int M = sc.nextInt(); // 연결고리의 갯수
        int maxCnt = 0;
        ArrayList<Integer> maxNodes = new ArrayList<>();

        // 2D 리스트 만들기
        ArrayList<ArrayList<Integer>> AL = new ArrayList<>();
        // 리스트 안에 빈 리스트 채우기
        for (int i=0; i<=N; i++) {
            AL.add(new ArrayList<>());
        }
        // 관계를 AL에 표시하기
        // A가 --> B를 신뢰하는 상황이라면 B를 해킹했을때 --> A도 해킹할 수 있음
        // AL[B]에 A를 추가
        for (int i=0; i<M; i++) {
            int A = sc.nextInt();
            int B = sc.nextInt();
            AL.get(B).add(A);
        }

        // 시작점을 1~N으로 설정해서 각각 DFS를 돈다
        for (int s=1; s<N; s++) {
            // java에서 int array는 자동으로 0으로 초기화된다
            int[] visited = new int[N+1];
            visited[s] = 1;
            ArrayList<Integer> stack = new ArrayList<>();
            stack.add(s);
            int cnt = 1;

            while (stack.isEmpty() == false) {
                int current = stack.remove(stack.size()-1);
                for (int i : AL.get(current)) {
                    if (visited[i] == 0) {
                        visited[i] = 1;
                        stack.add(i);
                        cnt++;
                    }
                }
            }

            if (cnt > maxCnt) {
                maxCnt = cnt;
                maxNodes = new ArrayList<>();
                maxNodes.add(s);
            } else if (cnt == maxCnt) {
                maxNodes.add(s);
            }
        }

        for (int i=0; i<maxNodes.size(); i++) {
            if (i == maxNodes.size()-1) {
                System.out.print(maxNodes.get(i));
            } else {
                System.out.print(maxNodes.get(i) + " ");
            }
        }
    }
}
